### ISSUE TITLE
Epic: Restart workspaces directly from VS Code Desktop

### ISSUE THREAD

**Issue by loujaybee** [gitpod-io/gitpod#9221]
### Press Release

⚠️ **To be confirmed (including user + customer quotes)**

### Context

_ℹ️ What thinking lead us to starting this work?_

VS Code Desktop is useful to some users because they can: 

- Re-use keybindings, (easier) access to localhost.
- Access to the extension marketplace.
- Desktop-like shortcuts (e.g. CMD + Tab to change windows). 

This is useful for users who don't like working in a browser, and want an experience that is already as close to their local experience as possible. However, the current VS Code Desktop has some rough edges, as compared with a simple desktop flow, particularly when it comes to opening (and then restarting a session). 

### Value

_ℹ️ What is the impact to a user, or Gitpod?_

We want to close the gap between the experience from working on VS Code Desktop in Gitpod, and the general local experience. We commonly hear from users that restarting sessions is quite challenging, as the user has to leave VS Code Desktop go to the browser, find the workspace, and restart. This is especially challenging if your workspace start time is long, or you have a lot of setup (e.g. other clients, like database, or SSH connections to re-establish). 

Restart an open workspace from VS Code Desktop directly. Currently timed out sessions have to be re-opened from the browser, however this should be as simple as clicking a button to restart / re-connect the session. 

![image](https://user-images.githubusercontent.com/5528307/181287362-88c936db-330d-4acc-871b-5962bcb179ad.png)

### Potential Solutions / Investigations

**Timebox the investigation:** Up until **Monday the 6th March.**

[Internal document on list of potential solutions.](https://www.notion.so/gitpod/VS-Code-Desktop-Restarting-cfed0de04de540bba730394909e56459?pvs=4)

### Acceptance Criteria 

⚠️ **Dependent on chosen solution**

### In Scope

⚠️ **Dependent on chosen solution**

### Out of scope

⚠️ **Dependent on chosen solution**

### Measurement

- Qualitative surveying / interviewing of the users experiences 
- No quantitative measures defined (to be confirmed)

**Relates to:** 
- https://github.com/gitpod-io/gitpod/issues/10798
- https://github.com/gitpod-io/gitpod/issues/15213
- https://github.com/gitpod-io/gitpod/issues/11335
- https://github.com/gitpod-io/gitpod/issues/10799
- https://github.com/gitpod-io/gitpod/issues/14280
- https://github.com/gitpod-io/gitpod/issues/10427
- https://github.com/gitpod-io/gitpod/issues/15419
- https://github.com/gitpod-io/gitpod/issues/15632
- https://github.com/gitpod-io/gitpod/issues/15829

**Comment by loujaybee** [comment]
https://github.com/gitpod-io/customers/issues/57

**Comment by mittalyashu** [comment]
Please add this option, it's really time consuming to open Gitpod dashboard, every-time the workspace times out.

Due to this limitation, I have to stick to single workspace use, for working on multiple tasks. Instead of using new workspace per task.

**Comment by loujaybee** [comment]
Scoping this issue down to only restart.

**Comment by loujaybee** [comment]
Comment(s) from user discussion ([internal link](https://gitpod.slack.com/archives/C01KPEPNBD0/p1671481349751709)): 

**User:** I missed a chance at recording it this morning, but described the flow below I'm talking about: Log on for the day, start up workspace for a PR I need to rebase main into. That goes fine, build is triggered. While I wait, I jump over to a PR to review. Come back to workspace I started ~ 15 minutes earlier to check something. I get the alert in vscode desktop that it's disconnected. I click retry, but that gives an error message w/ a few options, but none that I can find that would just restart the workspace. I end up with the editor in a disconnected state - and I don't see an option to start the workspace back up. Then I jump over to the browser, find my workspaces list on gitpod, and start the right one and wait for it again (sometimes that feels pretty quick, but sometimes it's several minutes). Another side effect is I end up with a bunch of abandoned vs code windows and gitpod tabs in the browser to maintain.

**Lou:** Would it be fair to summarise that the friction you’re talking about is mostly related to getting a VS Code Desktop workspace restarted?
1. Start a workspace
2. Workspace starting up
3. Gitpod workspace automation
4. Workspace times out
5. Find and restart workspace **<- friction 🎯**
6. Workspace starting up
7. Gitpod workspace automation

Step 5 reads like the most painful for you right now?

**User:** Yah, if 5 was simpler, that would help. The duration of waiting for 5 can be a big part of the friction though too (I think I waited ~ 8 minutes today for one to restart after timing out).

**Comment by jeanp413** [comment]
Related https://gitpod.slack.com/archives/C01KGM9BH54/p1676498720629379

**Comment by loujaybee** [comment]
I'm going to move this out of team IDE board for now, we'll create an issue for the proposed "local SSH solution" this solution _might_ fix issues with VS Code Desktop restarts (as originally intended in this issue) so then we can either pull this issue back in to the working stack or close it. Alternatively we can also pick this issue back up when the create workspace public API is working, as that should unblock us in implementing this. Either way this issue is not yet ready to be worked on, so removing from the current board for now.

**Comment by bradleybeighton** [comment]
I get the same error, is there an easy work around for this?

**Comment by mustard-mh** [comment]
> I get the same error, is there an easy work around for this?

Hey @bradleybeighton , which issue are you referring? This one https://github.com/gitpod-io/gitpod/issues/9221#issuecomment-1359706404 ?

**Comment by bradleybeighton** [comment]
@mustard-mh yes exactly, I can't re-connect to the gitpod instance when its disconnected using VS Code desktop

**Comment by mustard-mh** [comment]
@bradleybeighton Could you follow the [troubleshooting](https://www.gitpod.io/docs/help/troubleshooting#gitpod-logs-in-vs-code-browser-and-desktop) after you failed to re-connect to help us debug and understand your faced issue 🙏 ? (Workspace ID should be helpful too)
